core/src/main/java/org/apache/gravitino/authorization/PermissionManager.java

```diff
@@ -373,11 +377,302 @@ User revokeRolesFromUser(String metalake, List<String> roles, String user) {
+                    .withAuditInfo(auditInfo)
+                    .withSecurableObjects(updateSecurableObjects)
+                    .build();
+              });
+
+      // Execute the authorization plugin callback
+      authorizationPluginCallbackWrapper.execute();
+      return updatedRole;
+    } catch (NoSuchEntityException nse) {
+      LOG.error("Failed to grant, role {} does not exist in the metalake {}", role, metalake, nse);
+      throw new NoSuchRoleException(ROLE_DOES_NOT_EXIST_MSG, role, metalake);
+    } catch (IOException ioe) {
+      throw new RuntimeException(ioe);
+    }
+  }
+
+  private static SecurableObject updateGrantedSecurableObject(
+      String metalake,
+      String role,
+      MetadataObject object,
+      List<Privilege> privileges,
+      RoleEntity roleEntity,
+      SecurableObject targetObject,
+      AuthorizationPluginCallbackWrapper authorizationPluginCallbackWrapper) {
+    // Removed duplicated privileges by set
+    Set<Privilege> updatePrivileges = Sets.newHashSet();
+    updatePrivileges.addAll(targetObject.privileges());
+    // If old object contains all the privileges to grant, the object don't
+    // need to change.
+    if (updatePrivileges.containsAll(privileges)) {
+      return targetObject;
+    } else {
+      updatePrivileges.addAll(privileges);
+      AuthorizationUtils.checkDuplicatedNamePrivilege(privileges);
+
+      SecurableObject newSecurableObject =
+          SecurableObjects.parse(
+              targetObject.fullName(), targetObject.type(), Lists.newArrayList(updatePrivileges));
+
+      // We set authorization callback here, we won't execute this callback in this place.
+      // We will execute the callback after we execute the SQL transaction.
+      authorizationPluginCallbackWrapper.setCallBack(
+          () ->
+              AuthorizationUtils.callAuthorizationPluginForMetadataObject(
+                  metalake,
+                  object,
+                  authorizationPlugin -> {
+                    authorizationPlugin.onRoleUpdated(
+                        roleEntity,
+                        RoleChange.updateSecurableObject(role, targetObject, newSecurableObject));
+                  }));
+
+      return newSecurableObject;
+    }
+  }
+
+  private static SecurableObject createNewSecurableObject(
+      String metalake,
+      String role,
+      MetadataObject object,
+      List<Privilege> privileges,
+      RoleEntity roleEntity,
+      AuthorizationPluginCallbackWrapper authorizationPluginCallbackWrapper) {
+    // Add a new securable object if there not exists the object in the role
+    SecurableObject securableObject =
+        SecurableObjects.parse(object.fullName(), object.type(), Lists.newArrayList(privileges));
+
+    // We set authorization callback here, we won't execute this callback in this place.
+    // We will execute the callback after we execute the SQL transaction.
+    authorizationPluginCallbackWrapper.setCallBack(
```

Comment:
Callback is a word, please change to `Callback`.

Reply:
OK.